### ISSUE TITLE
Implementation of federated graphs

### ISSUE THREAD

**Issue by marcus-sa** [nestjs/graphql#299]
## I'm submitting a...

<pre><code>
[ ] Regression 
[ ] Bug report
[x] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

## Current behavior

No support for federated graphs

## Proposal

`@apollo/federation` uses a custom property in resolvers for referencing, just like when resolving a property.
Currently `@nestjs/graphql` have a `@ResolveProperty()` decorator, and what I propose is a `@ResolveReference()` decorator explicitly for when using `@apollo/federation` 

Example of how they use `__resolveReference`:
```ts
const { ApolloServer, gql } = require('apollo-server');

const { buildFederatedSchema } = require('@apollo/federation');

const typeDefs = gql`
  type Query {
    me: User
  }


  type User @key(fields: "id") {
    id: ID!
    username: String
  }
`;

const resolvers = {
  Query: {
    me() {
      return { id: "1", username: "@ava" }
    }
  },

  User: {
    __resolveReference(user, { fetchUserById }){
      return fetchUserById(user.id)
    }
  }
};

const server = new ApolloServer({

  schema: buildFederatedSchema([{ typeDefs, resolvers }])
});
``` 

## What is the motivation / use case for changing the behavior?

https://www.apollographql.com/docs/apollo-server/federation/introduction

## Links
https://www.apollographql.com/docs/apollo-server/api/apollo-federation/

**Comment by marcus-sa** [comment]
RFC: Should this be a part of the current `@nestjs/graphql` module, or should it be implemented as a community project?
Some people might not like the overheat that `@apollo/federation` provides if its useless for them.

**Comment by marcus-sa** [comment]
I have an example implementation of how it should be done in Nest (beware that this is just stitching together the explorer services)

https://github.com/marcus-sa/nest-graphql-microservices-example

**Comment by brianschardt** [comment]
Any good or official implementation of federation for nestjs. Perhaps it could be a config option so people can specify federated: true, default false. If they want to use federated graph.

**Comment by marcus-sa** [comment]
Haven't yet done any documentation, but essentially you'd be using `GraphQLGatewayModule` for your API gateway and `GraphQLFederationModule` for your microservices.

**Comment by brianschardt** [comment]
I see that makes sense. Anyone from nest looking into this. This is a super important feature!

**Comment by juicycleff** [comment]
Until a fix, you can use my library https://github.com/juicycleff/nestjs-graphql-gateway. You keep using this library for everything else except the module initialization

**Comment by marcus-sa** [comment]
@juicycleff where's the credits? You pretty much just copy-pasted my work.

**Comment by juicycleff** [comment]
@marcus-sa oh sorry about I added credits to. I basically used it in my project and decided to make it a lib so others can easily use it until there is support for federation here. Besides I was working on it, saw your library, wasn't much difference, took the part I needed and used it internally and decided to share it so others can use util support here, I even added support for typegraphql, while true I copied some piece of your work which you made public to copy for free, it was not COPY PASTE. However again my bad for not adding credits to you, besides it was just a temporary fix. and not too different from what I made before only my approach was almost replacing the nest graphql library
